ACE/ace/OS_NS_Thread.cpp

```diff
@@ -3807,6 +3990,72 @@ ACE_OS::thr_create (ACE_THR_FUNC func,
       return 0;
     }
 
+# elif defined (ACE_INTEGRITY)
+    ACE_UNUSED_ARG (flags);
+
+    // From INTEGRITY-178 Kernel Reference Guide, Sec. 4 - Task Kernel Calls for SetupTask:
+    // "Priority must be less than or equal to the Priority of the source Task. Moreover, when the head
+    // is being used (e.g., when using malloc or the C++ new operator), Priority must be set to the
+    // Maximum Priority of the Initial Task so that the Maximum Priority of the new Task will be the
+    // same as that of the Initial Task."
+    // We start from the Initial Task and all new Tasks will use the same Maximum Priority.
+    Value max_parent_prio, unused_weight;
+    Error err_code = ::GetMaximumPriorityAndWeight (CurrentTask (), &max_parent_prio, &unused_weight);
+    if (err_code != Success)
+      return -1;
+
+    // Disregard the input priority value
+    priority = max_parent_prio;
+
+#   if defined (ACE_INTEGRITY178B)
+    // Use the pre-allocated stack if no stack provided
+    unsigned slot = NUM_DYNAMIC_THREADS;
+    if (!stack)
+      {
+        stack = int178_stack_manager.acquire (slot);
+        if (!stack)
+          return -1;
+        stacksize = ACE_INTEGRITY178B_DEFAULT_STACK_SIZE;
+      }
+
+    // Don't let this Task run immediately; we have to set its entrypoint's argument below.
+    err_code = ::SetupTask (AddressSpaceObjectNumber (1), priority /* priority */, 0 /* weight */, 0 /* timeslice */,
+                          (Address) stack /* stack */, (Address) stacksize /* stackLength */, true /* enableClibrary */,
+                          true /* allocTLS */, (Address) &integrity_task_adapter, false /* startIt */, 0 /* name */,
+                          0 /* symbolFile */, thr_handle /* newTask */, 0 /* newActivity */);
+
+    // Set the owner of the acquired stack if one from the stack pool is being used.
+    if (slot != NUM_DYNAMIC_THREADS)
+      {
+        int178_stack_manager.owner (*thr_handle, slot);
+      }
+
+#   else
+    ACE_UNUSED_ARG (stack);
+    // INTEGRITY has kernel calls in which we can pass argument for the new Task's entry point,
+    // e.g. CommonCreateTaskWithArgument.
+    // However, we are imitating INTEGRITY-178 and the call without entrypoint's argument is used.
+    err_code = ::CommonCreateTask (priority, (Address) &integrity_task_adapter /* entrypoint */,
+                                   (Address) stacksize, 0 /* name */, thr_handle /* newtask */);
+#   endif
+
+    if (err_code != Success)
+      return -1;
+
+    *thr_id = *thr_handle;
+
+    // Set the argument for the new Task's entrypoint, then run it.
+    mutex_lock (&integrity_task_args_lock);
+    integrity_task_args[*thr_handle] = thread_args;
+
+    Semaphore sema;
+    if (::CreateSemaphore (0, &sema) != Success)
+      return -1;
```

Comment:
You are not releasing integrity_task_args_lock when this fails, maybe use a guard for all calls to mutex_lock/mutex_unlock?

Reply:
Good catch. Thanks!